drivers/spi/spi_ll_stm32.c

```diff
@@ -468,129 +503,60 @@ static int spi_stm32_init(struct device *dev)
-	SPI_CONTEXT_INIT_SYNC(spi_stm32_dev_data_1, ctx),
-};
-
-DEVICE_AND_API_INIT(spi_stm32_1, CONFIG_SPI_1_NAME, &spi_stm32_init,
-		    &spi_stm32_dev_data_1, &spi_stm32_cfg_1,
-		    POST_KERNEL, CONFIG_SPI_INIT_PRIORITY,
-		    &api_funcs);
-
-#ifdef CONFIG_SPI_STM32_INTERRUPT
-static void spi_stm32_irq_config_func_1(struct device *dev)
-{
-	IRQ_CONNECT(CONFIG_SPI_1_IRQ, CONFIG_SPI_1_IRQ_PRI,
-		    spi_stm32_isr, DEVICE_GET(spi_stm32_1), 0);
-	irq_enable(CONFIG_SPI_1_IRQ);
-}
-#endif
-
-#endif /* CONFIG_SPI_1 */
-
-#ifdef CONFIG_SPI_2
-
-#ifdef CONFIG_SPI_STM32_INTERRUPT
-static void spi_stm32_irq_config_func_2(struct device *port);
-#endif
-
-static const struct spi_stm32_config spi_stm32_cfg_2 = {
-	.spi = (SPI_TypeDef *) CONFIG_SPI_2_BASE_ADDRESS,
-	.pclken = {
-		.enr = LL_APB1_GRP1_PERIPH_SPI2,
-		.bus = STM32_CLOCK_BUS_APB1
-	},
-#ifdef CONFIG_SPI_STM32_INTERRUPT
-	.irq_config = spi_stm32_irq_config_func_2,
-#endif
-};
-
-static struct spi_stm32_data spi_stm32_dev_data_2 = {
-	SPI_CONTEXT_INIT_LOCK(spi_stm32_dev_data_2, ctx),
-	SPI_CONTEXT_INIT_SYNC(spi_stm32_dev_data_2, ctx),
-};
-
-DEVICE_AND_API_INIT(spi_stm32_2, CONFIG_SPI_2_NAME, &spi_stm32_init,
-		    &spi_stm32_dev_data_2, &spi_stm32_cfg_2,
-		    POST_KERNEL, CONFIG_SPI_INIT_PRIORITY,
-		    &api_funcs);
-
-#ifdef CONFIG_SPI_STM32_INTERRUPT
-static void spi_stm32_irq_config_func_2(struct device *dev)
-{
-	IRQ_CONNECT(CONFIG_SPI_2_IRQ, CONFIG_SPI_2_IRQ_PRI,
-		    spi_stm32_isr, DEVICE_GET(spi_stm32_2), 0);
-	irq_enable(CONFIG_SPI_2_IRQ);
-}
-#endif
-
-#endif /* CONFIG_SPI_2 */
-
-#ifdef CONFIG_SPI_3
-
-#ifdef CONFIG_SPI_STM32_INTERRUPT
-static void spi_stm32_irq_config_func_3(struct device *port);
-#endif
-
-static const  struct spi_stm32_config spi_stm32_cfg_3 = {
-	.spi = (SPI_TypeDef *) CONFIG_SPI_3_BASE_ADDRESS,
-	.pclken = {
-		.enr = LL_APB1_GRP1_PERIPH_SPI3,
-		.bus = STM32_CLOCK_BUS_APB1
-	},
-#ifdef CONFIG_SPI_STM32_INTERRUPT
-	.irq_config = spi_stm32_irq_config_func_3,
-#endif
-};
-
-static struct spi_stm32_data spi_stm32_dev_data_3 = {
-	SPI_CONTEXT_INIT_LOCK(spi_stm32_dev_data_3, ctx),
-	SPI_CONTEXT_INIT_SYNC(spi_stm32_dev_data_3, ctx),
-};
-
-DEVICE_AND_API_INIT(spi_stm32_3, CONFIG_SPI_3_NAME, &spi_stm32_init,
-		    &spi_stm32_dev_data_3, &spi_stm32_cfg_3,
-		    POST_KERNEL, CONFIG_SPI_INIT_PRIORITY,
-		    &api_funcs);
-
-#ifdef CONFIG_SPI_STM32_INTERRUPT
-static void spi_stm32_irq_config_func_3(struct device *dev)
-{
-	IRQ_CONNECT(CONFIG_SPI_3_IRQ, CONFIG_SPI_3_IRQ_PRI,
-		    spi_stm32_isr, DEVICE_GET(spi_stm32_3), 0);
-	irq_enable(CONFIG_SPI_3_IRQ);
-}
-#endif
+/**
+ * @code{.codegen}
+ * codegen.import_module('devicedeclare')
+ *
+ * device_configs = ['CONFIG_SPI_{}'.format(x) for x in range(1, 7)]
+ * driver_names = ['SPI_{}'.format(x) for x in range(1, 7)]
+ * device_inits = 'spi_stm32_init'
+ * device_levels = 'POST_KERNEL'
+ * device_prios = 'CONFIG_SPI_INIT_PRIORITY'
+ * device_api = 'spi_stm32_driver_api'
+ * device_info = \
```

Comment:
I maybe missed a bit, but where do you declare the device compatible ? How do you link with the DT node and it's attributes ?

Reply:
I declare the driver names. Device compatible is then read from the device tree.

Reply:
Could you clarify link between driver names and DT?

Reply:
The driver name (aka. the label property from DTS) is used to identify the driver instance and to retrieve all the data from the extended DTS database.

Reply:
Is it a Zephyr policy to use label for that ?
In the normal DT world (Linux, U-Boot, ...) we use the compatible and the aliases to determine the driver and the id of the device, the "label" property is only used to give a label on the device, not to link device<->driver.
AFAIK we used the compatible to generate the defines  (ST_STM32_I2C_V2_*) and we used the label to generate the config (CONFIG_I2C_1_*). Then we used the fixup to link the two.
Why can't we generate the fixups using codegen ?

Reply:
@b0661 , I agree with @superna9999  on the usual use of 'compatible'. It should be sufficient to link driver to the nodes. Using 'label', you might face issues, for instance when it is not numbered, but use letters for enumeration, or with only one instance.

@superna9999 , indeed they can be generated. But ideally fixups should not be needed and we should be able to generate directly the device init code with values from dt, w/o using intermediate defines. Btw, in #8561, I'm demonstrating an generation API, also based on codegen, and using 'compatible' as the link between driver and dts file.

Reply:
@erwango 

> Using 'label', you might face issues, for instance when it is not numbered, but use letters for enumeration, or with only one instance.

This is not a problem as the driver writer knows about the labels and can create whatever list of labels is necessary. This is not restricted to numbers.

The driver writer can also derive the labels from the 'compatible' devices as you do in your device_declare API. As this is a raw API it is just made visible. The real benefit is that you do not have to assume/ impose that all config values are of the form CONFIG_[label] but can be of any form as long as they are listen in the same sequence as the related labels.

You always need a link between the driver instance and the DTS instance. There are only two practical properties, the label and the base address.  Also in your device_declare API you have choosen to link the DTS by label. It is just hidden behind the API. The label is first searched by the 'compatible' property and then used for the CONFIG_[label] value.

To have full flexibilty for the raw version I am using the label for the linkage to the DTS. How these labels are found/ defined is up to a more sophisticated API.